File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/voicechat/VarInt.java
package crabcraft.net.crabUtilities.voicechat;

import java.io.ByteArrayOutputStream;

/**
 * Minecraft VarInt encoder. We need this for the {@code String} field
 * inside SVC's {@code PlayerStatePacket} wire format, which is encoded
 * as a VarInt-length-prefix followed by UTF-8 bytes (the same encoding
 * Mojang's {@code FriendlyByteBuf.writeUtf} uses).
 */
final class VarInt {

    private VarInt() {}

    static void write(ByteArrayOutputStream out, int value) {
        while ((value & ~0x7F) != 0) {
            out.write((value & 0x7F) | 0x80);
            value >>>= 7;
        }
        out.write(value & 0x7F);
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/voicechat/VarInt.kt
package crabcraft.net.crabUtilities.voicechat

import java.io.ByteArrayOutputStream

/** Minecraft VarInt encoding for SVC's UTF-8 PlayerStatePacket fields. */
object VarInt {
    @JvmStatic
    fun write(out: ByteArrayOutputStream, value: Int) {
        var remaining = value
        while (remaining and 0x7F.inv() != 0) {
            out.write((remaining and 0x7F) or 0x80)
            remaining = remaining ushr 7
        }
        out.write(remaining and 0x7F)
    }
}
